Source/WebCore/dom/Element.cpp

```diff
@@ -3296,6 +3328,26 @@ RefPtr<ShadowRoot> Element::shadowRootForBindings(JSC::JSGlobalObject& lexicalGl
     return nullptr;
 }
 
+RefPtr<Element> Element::deepShadowRootReferenceTargetOrSelf()
+{
+    return const_cast<Element*>(const_cast<const Element*>(this)->deepShadowRootReferenceTargetOrSelf().get());
+}
+
+RefPtr<const Element> Element::deepShadowRootReferenceTargetOrSelf() const
+{
+    if (!document().settings().shadowRootReferenceTargetEnabled())
+        return this;
+
+    RefPtr element = this;
+    RefPtr shadow = shadowRoot();
+
+    while (shadow && shadow->hasReferenceTarget()) {
+        element = shadow->referenceTargetElementOrHost();
```

Comment:
It's odd that this function falls back to the host when reference target isn't set after we've checked that the shadow root has a reference target. Why not just add a member function which returns `referenceTargetElement` or `nullptr`?